nomad/state/state_store_test.go

```diff
@@ -7678,232 +7685,228 @@ func TestStateStore_SetJobStatus(t *testing.T) {
-	if err != nil {
-		t.Fatalf("getJobStatus() failed: %v", err)
-	}
+				a.ClientStatus = structs.AllocClientStatusFailed
+				a.JobID = j.ID
 
-	if status != structs.JobStatusPending {
-		t.Fatalf("getJobStatus() returned %v; expected %v", status, structs.JobStatusPending)
-	}
-}
+				e := mock.Eval()
+				e.JobID = j.ID
+				e.JobModifyIndex = j.ModifyIndex
+				e.Status = structs.EvalStatusPending
 
-// TestStateStore_SetJobStatus_SystemJob asserts that system jobs are still
-// considered running until explicitly stopped.
-func TestStateStore_SetJobStatus_SystemJob(t *testing.T) {
-	ci.Parallel(t)
+				if err := txn.Insert("allocs", a); err != nil {
+					return nil, err
+				}
 
-	state := testStateStore(t)
-	job := mock.SystemJob()
+				if err := txn.Insert("evals", e); err != nil {
+					return nil, err
+				}
+				return j, nil
 
-	// Create a mock eval that is pending.
-	eval := mock.Eval()
-	eval.JobID = job.ID
-	eval.Type = job.Type
-	eval.Status = structs.EvalStatusComplete
-	if err := state.UpsertEvals(structs.MsgTypeTestSetup, 1000, []*structs.Evaluation{eval}); err != nil {
-		t.Fatalf("err: %v", err)
+			},
+			exp: structs.JobStatusPending,
+		},
+		{
+			name: "reschedulable alloc is pending waiting for replacement",
+			setup: func(t *txn) (*structs.Job, error) {
+				j := mock.Job()
+				if j.TaskGroups[0].ReschedulePolicy == nil {
+					return nil, errors.New("mock job doesn't have replacement policy")
+				}
+				a := mock.Alloc()
+				a.Job = j
+				a.JobID = j.ID
+				a.ClientStatus = structs.AllocClientStatusFailed
+				if err := t.Insert("allocs", a); err != nil {
+					return nil, err
+				}
+				return j, nil
+			},
+			exp: structs.JobStatusPending,
+		},
+		{
+			name: "reschedulable alloc is dead after replacement fails",
+			setup: func(t *txn) (*structs.Job, error) {
+				j := mock.Job()
+				// give job one reschedule attempt
+				j.TaskGroups[0].ReschedulePolicy.Attempts = 1
+				j.TaskGroups[0].ReschedulePolicy.Interval = time.Hour
+
+				// Replacement alloc
+				a := mock.Alloc()
+				a.Job = j
+				a.JobID = j.ID
+				a.ClientStatus = structs.AllocClientStatusFailed
+				a.RescheduleTracker = &structs.RescheduleTracker{
+					Events: []*structs.RescheduleEvent{
+						structs.NewRescheduleEvent(time.Now().UTC().UnixNano(), "", "", time.Minute),
+					},
+				}
+				if err := t.Insert("allocs", a); err != nil {
+					return nil, err
+				}
+
+				// Original alloc
+				a2 := mock.Alloc()
+				a2.Job = j
+				a2.JobID = j.ID
+				a2.ClientStatus = structs.AllocClientStatusFailed
+				a2.NextAllocation = a.ID
+				if err := t.Insert("allocs", a2); err != nil {
+					return nil, err
+				}
+				return j, nil
+			},
+			exp: structs.JobStatusDead,
+		},
 	}
 
-	txn := state.db.ReadTxn()
-	status, err := state.getJobStatus(txn, job, true)
-	if err != nil {
-		t.Fatalf("getJobStatus() failed: %v", err)
-	}
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
```

Comment:
We can probably run this inner loop in parallel. Doesn't look like there's anything shared between tests.